### PR TITLE
Update C++ samples

### PR DESCRIPTION
**Pull request by akiramenai**
* Add C++ Tutorial.
* Add Hello, world sample.
* Add Authorization sample.
* Add Kamikaze sample.

### CODE REVIEW

File: cpp/README.md [hunk 1]
```diff
@@ -43,12 +43,11 @@ If the compilation is successful, `*.tvc` file is created in the directory where
...
 ## Contract deployment
 
-Here we describe a contract deployment to the TON Blockchain Test Network (testnet) using `tonlabs-cli` tool.
-Note: for MacOS X replace `tonlabs-cli` with `cargo run` in the following instruction, and the run is supposed to be made from the directory where `tonlabs-cli` is built.
+Here we describe a contract deployment to the TON Blockchain Test Network (testnet) using `tonlabs-cli` tool. The following commands must be run from the directory where tonlabs-cli is located.
 
 1. Generate contract address
 ```
-tonlabs-cli genaddr Contract.tvc Contract.abi --genkey key
+cargo run Contract.tvc Contract.abi --genkey key
```
Comment (azhogin) [inline cpp/README.md:50]: Are you sure that you don't need `genaddr` or other command here?

File: cpp/README.md [hunk 2]
```diff
@@ -43,12 +43,11 @@ If the compilation is successful, `*.tvc` file is created in the directory where
 
 ## Contract deployment
 
-Here we describe a contract deployment to the TON Blockchain Test Network (testnet) using `tonlabs-cli` tool.
-Note: for MacOS X replace `tonlabs-cli` with `cargo run` in the following instruction, and the run is supposed to be made from the directory where `tonlabs-cli` is built.
+Here we describe a contract deployment to the TON Blockchain Test Network (testnet) using `tonlabs-cli` tool. The following commands must be run from the directory where tonlabs-cli is located.
```
Comment (azhogin) [inline cpp/README.md:46]: I've heard that tonlabs-cli is renamed to tonos-cli (https://github.com/tonlabs/tonos-cli).
Should we continue to keep old name in docs?

File: cpp/TUTORIAL.md [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+To minimize the state which is transferred between contracts and stored in the blockchain, TVM does not have random access memory. The memory might be emulated by dictionaries, however, it’s quite expensive and often not practical taking into account limited gas supply. Thus a typical smart contract in C++ avoid use memory. Moreover, currently all the functions have to be inlined. That’s why `-O3` optimization and LTO are essential and it’s the pipeline `tvm-build++` implements.
+
+## Environment setup
+Prior to start developing a contract, we configure `PATH` to add all the tools we need:
+```
+export PATH=TON-Compiler-build/bin:TVM-linker-source/tvm_linker/target/<debug or release>:TVM-linker-source/tonlabs-cli/target/<debug or release>:$PATH
+```
+`tvm-build++` needs additional theatment:
```
Comment (azhogin) [inline cpp/TUTORIAL.md:27]: `treatment` maybe?

File: cpp/TUTORIAL.md [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+*	Public methods declarations.
+*	Persistent data.
+*	Events.
+Public methods are functions that receive messages in the blockchain. There is a dedicated method called `constructor` which is called upon the contract deploy. The constructor is a must otherwise the contract can not be deployed. The constructor, as well as other public methods, must only take arguments of types that are listed below:
+*	`int_t<N>` - N-bits wide signed integer, 0 < N < 256.
+*	`uint_t<N>` - N-bits wide unsigned integer, 0 < N < 257.
+*	`MsgAddress` – message address which work for both internal or external messages.
+*	`MsgAddressInt` – an internal message address.
+*	`MsgAddressExt` – an external message address.
+*	`dict_array<T, 32>` - a map representing an "array" with up to 32 elements of type `T` (indexed from 0 to 31). `T` must also belong to this list.
```
Comment (azhogin) [inline cpp/TUTORIAL.md:53]: 32 is not elements count in dict_array, it is keylen in bits.
```
template<class Element, unsigned KeyLen = 32>
class dict_array 
```

File: cpp/TUTORIAL.md [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+*	`dict_map<KeyT, ValueT>` - a map. Both `KeyT` and `ValueT` must also belong to the list.
```
Comment (azhogin) [inline cpp/TUTORIAL.md:54]: Actually, KeyT must be `uint_t<N>` strictly (in dict_map).

File: cpp/TUTORIAL.md [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+*	`sequence<uint_t<8>>` - a sequence of 8-bit unsigned values, it might also be seen as an "array" of 8-bits values; it’s usually cheaper to use sequentially stored data then maps.
+*	`lazy<T>` - value of `T`, but the parsing is deferred until the actual value is needed.
```
Comment (azhogin) [inline cpp/TUTORIAL.md:56]: Maybe add phrase: Only `lazy<MsgAddress>`, `lazy<MsgAddressInt>` and `lazy<MsgAddressExt>` are supported atm.

File: cpp/README.md (hunk 2 above)
Comment (akiramenai) [reply cpp/README.md:46]: `tonlabs-cli` README https://github.com/tonlabs/TVM-linker/tree/master/tonlabs-cli still uses `tonlabs-cli` so I stick with it. When the tool is renamed, I update the docs.

File: cpp/README.md (hunk 1 above)
Comment (akiramenai) [reply cpp/README.md]: Fixed

File: cpp/TUTORIAL.md (hunk 3 above)
Comment (akiramenai) [reply cpp/TUTORIAL.md]: Fixed

File: cpp/TUTORIAL.md (hunk 4 above)
Comment (akiramenai) [reply cpp/TUTORIAL.md]: Fixed

File: cpp/TUTORIAL.md (hunk 5 above)
Comment (akiramenai) [reply cpp/TUTORIAL.md]: Fixed

File: cpp/TUTORIAL.md (hunk 6 above)
Comment (akiramenai) [reply cpp/TUTORIAL.md]: Added

**Pull request merged**